### DISCUSSION TITLE
creating-a-source-generator-part-10-testing-your-incremental-generator-pipeline-outputs-are-cacheable/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#406] [Announcements]
# creating-a-source-generator-part-10-testing-your-incremental-generator-pipeline-outputs-are-cacheable/

In this post I show an approach for unit testing that your incremental generator outputs are cached correctly so you don't impact IDE performance

https://andrewlock.net/creating-a-source-generator-part-10-testing-your-incremental-generator-pipeline-outputs-are-cacheable/

**Comment by JasonBock** [comment]
In the first section, you say, "It uses `ForAttributeWithMetadataName`", but your code example is calling `CreateSyntaxProvider()`.

**Reply by andrewlock** [reply]
🤦‍♂️ Oops, thanks.

**Comment by OlegKarasik** [comment]
Very good addition to the series. 

I think I have found a typo. In the “Adding tracking names to your stages” section you are defining two tack names but in the following code only “InitialExtraction” one is used. Probably the second call to “WithTrackingName” must use “Transform” instead of “InitialExtraction”.

**Comment by kzu** [comment]
A deeper look at the [Roslyn implementation of the `ForAttributeWithMetadataName`](https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/SourceGeneration/Nodes/SyntaxValueProvider_ForAttributeWithSimpleName.cs#L55) recommended method showcases that the overall post recommendations are actually going way too far. 

You can perfectly well combine your sources with the compilation provider, and still have a fine generator. I think the underlying recommendation should be "avoid the syntax provider like the plague" and that's it. Also not caching symbol or compilation data. But that's about it, IMHO. (that's what [STJ generator checks](https://github.com/dotnet/runtime/blob/b4488eca7f87cde6f2a19d388b79a723d9ec71e8/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Unit.Tests/JsonSourceGeneratorIncrementalTests.cs#L148C13-L183C18))

**Comment by cbersch** [comment]
To check for unallowed types with `FluentAssertions` or `AwesomeAssertions`, use `NotBeAssignableTo`, because `NotBeOfType` only checks for type equality.

**Reply by bkoelman** [reply]
I have an incremental generator that consists of multiple stages. The first stage performs basic analysis. The next stage enriches the gathered information by making some expensive calls, which are needed in subsequent stages. I'm using multiple stages, so that the subsequent logic doesn't need to execute when the source code has changed in a way that doesn't change the gathered basic information.

The test code presented here only verifies that the outputs of the _first_ stage are cacheable. Because the test analyzes the same source twice, the first stage is always cached when the second run executes. So any later stages don't execute during the second run, therefore non-cacheable outputs in them aren't reported. Or am I missing something?

**Reply by bkoelman** [reply]
Oh, I didn't intend to reply to an existing comment. Will re-post as top-level (please delete this reply).

**Comment by bkoelman** [comment]
I have an incremental generator that consists of multiple stages. The first stage performs basic analysis. The next stage enriches the gathered information by making some expensive calls, which are needed in subsequent stages. I'm using multiple stages, so that the subsequent logic doesn't need to execute when the source code has changed in a way that doesn't change the gathered basic information.

The test code presented here only verifies that the outputs of the _first_ stage are cacheable. Because the test analyzes the same source twice, the first stage is always cached when the second run executes. So any later stages don't execute during the second run, therefore non-cacheable outputs in them aren't reported. Or am I missing something?
